rules/when-to-use-domain-and-integration-events/rule.md

```diff
@@ -0,0 +1,61 @@
+---
+type: rule
+title: Do you know when to use Domain and Integration Events?
+guid: e91f22d2-d385-49b6-a27b-1195b4bb13f9
+uri: when-to-use-domain-and-integration-events
+created: 2023-07-25T11:01:35.0000000Z
+authors:
+- title: William Liebenberg
+  url: https://ssw.com.au/people/william-liebenberg
+archivedreason:
+related: []
+redirects:
+---
+
+Domain Events and Integration Events are powerful patterns used in Domain-Driven Design (DDD) to enable loosely coupled communication between different parts of an application. They serve as a means of **notification** for important domain concepts that have occurred.
+
+<!--endintro-->
+
+## What are Domain Events
+
+In the Domain Layer of the [Clean Architecture](https://github.com/sswconsulting/SSW.CleanArchitecture), Domain Events are an integral part of the Domain Model.
+
+Domain Events are raised within the Domain Layer of your application when an entity or aggregate makes a significant decision or undergoes a state change. The main purpose of Domain Events is to enable loose coupling and keep domain logic isolated from the application's infrastructure.
+
+For example, when an `Order` is placed, a domain event can be raised (e.g. `OrderReceivedEvent`) to notify other parts of the domain that need to react to this event, such as updating inventory or sending confirmation emails.
+
+Another example could be when the `Order`'s `Status` changes from `OrderStatus.Received` to `OrderStatus.Processing`, we can publish an `OrderStatusChanged` event.
+
+It's essential to identify suitable events as not everything qualifies as a Domain Event. For instance, "client walked into the store" or "chicken crossed the road" may not be appropriate for a Domain Event.
+
+Domain events are immutable and can be considered as **historical facts** capturing something that occurred in the domain process. They are meant to be a representation of past events and cannot be altered or disputed.
+
+When describing domain events, we commonly use a past-tense naming convention, such as `OrderCreated`, `UserRegistered`, `InvoiceConsolidated`.
+
+Domain events should not depend on external dependencies or external systems, adhering to the principles of the Domain Layer.
+
+When publishing a domain event, the entire Entity or Aggregate can be included since the event's scope is confined to the current bounded context.
+
+Note: It is important to remember that the definition and behavior of a `Product` in one bounded context might differ from another bounded context, like an e-commerce application product versus the product of a chemical reaction in a laboratory.
+
+In the `Application` Layer, domain events are typically handled synchronously. Any database side-effects are tracked as part of the current transaction of the original request, ensuring strong consistency in the response sent back to the user once the transaction is committed.
```

Comment:
Didn't we work out in our chat last week that this isn't the approach to use?

> [18/07 5:30 pm] Daniel Mackay [SSW]
> 
> TLDR; should work fine, but be aware of the differences between using Find() vs something like ToListAsync()
> 
> [18/07 5:32 pm] Matt Goldman [SSW]
> 
> Also use Saved I think?
> 
> 
> [18/07 5:33 pm] Matt Goldman [SSW]
> 
> So it should publish the events after the save is completed, and any additional changes should be a new query
> 
> 
> [18/07 5:34 pm] William Liebenberg [SSW]
> 
> yea...multiple sequential transactipons
> 
> [18/07 5:36 pm] Daniel Mackay [SSW]
> 
> Thinking about this more, it makes sense.  
> 
>  * Command updates domain
>  * Events fired
>  * Event Handlers triggered
>  * More domain updates
>  * More events fired (possibly)
>  * More event handlers triggered (possibly)
>  * (and so on and so forth)
> 
>  
> 
> [18/07 5:37 pm] Daniel Mackay [SSW]
> 
> And watch out for any infinite loops!


I think this probably needs its own rule. And I'm sure @danielmackay is keen to write `Rules to better DDD` anyway :)

Reply:
Or...maybe I keep reading and the process in that chat is for integration events...?

Reply:
yea the Domain event side-effects can occur is in the original transaction (Saving) like we have it now. Integration events should happen after the transaction is committed and asynchronously as well so that we don't block the application flow/user request. I'll have these examples in the demo app.